clash-lib/src/Clash/Driver/Types.hs

```diff
@@ -78,25 +80,172 @@ data Binding a = Binding
+  | FinalTerm
+  -- ^ Show the final term after all applied transformations.
+  | AppliedName
+  -- ^ Show the name of every transformation that is applied.
+  | AppliedTerm
+  -- ^ Show the name and result of every transformation that is applied.
+  | TryName
+  -- ^ Show the name of every transformation that is attempted, and the result
+  -- of every transformation that is applied.
+  | TryTerm
+  -- ^ Show the name and input to every transformation that is applied, and
+  -- the result of every transformation that is applied.
+  deriving (Eq, Generic, Hashable, Ord, Read, Show)
+
+-- | Options related to debugging. See 'ClashOpts'
+data DebugOpts = DebugOpts
+  { opt_invariants :: Bool
+  -- ^ Check that the results of applied transformations do not violate the
+  -- invariants for rewriting (e.g. no accidental shadowing, or type changes).
+  --
+  -- Command line flag: -fclash-debug-invariants
+  , opt_transformationInfo :: TransformationInfo
+  -- ^ The information to show when debugging a transformation. See the
+  -- 'TransformationInfo' type for different configurations.
+  --
+  -- Command line flag: -fclash-debug-info (None|FinalTerm|AppliedName|AppliedTerm|TryName|TryTerm)
+  , opt_transformations :: Set String
+  -- ^ List the transformations that are being debugged. When the set is empty,
+  -- all transformations are debugged.
+  --
+  -- Command line flag: -fclash-debug-transformations t1[,t2...]
+  , opt_countTransformations :: Bool
+  -- ^ Count how many times transformations are applied and provide a summary
+  -- at the end of normalization. This includes all transformations, not just
+  -- those in 'opt_transformations'.
+  --
+  -- Command line flag: -fclash-debug-count
+  , opt_transformationsFrom :: Maybe Word
+  -- ^ Debug transformations applied after the nth transformation applied. This
+  -- includes all transformations, not just those in 'opt_transformations'.
+  --
+  -- Command line flag: -fclash-debug-transformations-from=N
+  , opt_transformationsLimit :: Maybe Word
+  -- ^ Debug up to the nth applied transformation. If this limit is exceeded
+  -- then Clash will error. This includes all transformations, not just those
+  -- in 'opt_transformations'.
+  --
+  -- Command line flag: -fclash-debug-transformations-limit=N
+  , opt_historyFile :: Maybe FilePath
+  -- ^ Save information about all applied transformations to a history file
+  -- for use with @clash-term@.
+  --
+  -- Command line flag: -fclash-debug-history[=FILE]
+  } deriving (Show)
+
+instance Hashable DebugOpts where
+  hashWithSalt s DebugOpts{..} =
+    s `hashWithSalt`
+    opt_invariants `hashWithSalt`
+    opt_transformationInfo `hashSet`
+    opt_transformations `hashWithSalt`
+    opt_countTransformations `hashWithSalt`
+    opt_transformationsFrom `hashWithSalt`
+    opt_transformationsLimit `hashWithSalt`
+    opt_historyFile
+   where
+    hashSet = Set.foldl' hashWithSalt
+    infixl 0 `hashSet`
```

Comment:
Any reason to not just derive `Hashable`?

Reply:
I did wonder about that. `Hashable` isn't derived for `ClashOpts`, so when I defined this one I just moved the relevant lines from that instance + added the ones for the new fields to be safe

Reply:
Hmmm, my guess is that that's a leftover from a feature/hack we've since moved to the right place?:

https://github.com/clash-lang/clash-compiler/blob/aaf15590d74b9477cfca96c0ab40e27e1b678ab3/clash-lib/src/Clash/Driver/Manifest.hs#L370-L408

I think we can just remove it and derive `Hashable` - in a separate commit of course so we can rollback if necessary.

Reply:
Ah, `Set String` has no `Hashable` instance, so deriving would fail

Reply:
And same for `OverridingBool` in `ClashOpts`. So I guess without making any more `*.Extra` modules we wouldn't be able to derive these instances

Reply:
Ah. Do we rely on the ordering? `HashSet` is generally preferable anyway. Which would leave `OverridingBool` - I guess an orphan instance wouldn't hurt.

Reply:
> Do we rely on the ordering?

No, it's just a set of transformations which are whitelisted for showing in the debug logs. I had thought that there should probably be a later commit to change it from `String` to `Text`, maybe that should be a commit to change it from `Set String` to `HashSet Text`?

Reply:
Yesss :)

Reply:
I think I'm going to save this kind of type shake-up for a future PR, there are a few places where the types in rewrite / normalize could be improved and fiddling with types seems a bit tangential to this PR